### DISCUSSION TITLE
How to handle when keys expire at clients end (in exe build app)

### DISCUSSION THREAD

**Discussion by ritishbhardwaj** [dennisvang/tufup#134] [Q&A]
for example i have given expiry as 3 days to timestamp and build an application at 1.0.0 and give it to user,  now i have updated my app to 1.0.1 on server after 5 days and now when user (1.0.0) click for an update it is unable to update and it gets keys expired.   How to handle this? How to refresh client's app so that keys gets refreshed and update can happen.

**Comment by dennisvang** [comment]
@ritishbhardwaj This needs to be handled on the *server* side.

If you specify `timestamp` expiry `n` days, then you need to re-sign the timestamp before those `n` days are over. 

This re-signing needs to be repeated every `n` days, indefinitely, so it's easiest to do it directly on your server, e.g. using a cron job.

For testing and development, you can do it locally and then upload the newly signed file to your server.

Once the file has been re-signed, any client that checks the file will proceed normally.

Also see this related discussion: #95

**Reply by ritishbhardwaj** [reply]
Ok, I have understood this. But still i have few questions where i need some help.

First:  you are saying to re-sign the the timestamp,i want to ask that what is the way to re-sign it, whether do we need to run the tufup repository init script again [https://github.com/dennisvang/tufup-example/blob/master/repo_init.py](url) so that the tufup will change the metadata or we have to add the signature manually using the cryptographic algorithms like RSA and make changes inside the Json metadata or when i add pyinstaller .exe bundle to the repository the re-signing will occur accordingly/automatically.

Second: with your answer i am coming up with one more doubt like suppose in the client machine where application is running and it has root.json file along with it [our main.spec ensures that the latest root.json metadata file is included in the bundle (given in example)] and when we re-sign the json file , eventually/obviously the changes will be made on the server side tufup repository's metadata files, then how would the client side signature will match the changed signature when the Client downloads the metadata over the client machine, it will eventually rejects the new archive and shows Error.

**Comment by dennisvang** [comment]
> First: you are saying to re-sign the the timestamp,i want to ask that what is the way to re-sign it, whether do we need to run the tufup repository init script again [https://github.com/dennisvang/tufup-example/blob/master/repo_init.py](url) so that the tufup will change the metadata or we have to add the signature manually using the cryptographic algorithms like RSA inside the Json metadata or when i add pyinstaller .exe bundle to the repository the re-signing changes accordingly/automatically.

@ritishbhardwaj The `repo_init.py` script (or `tufup init` on the command line) should only be used *once*, for *initialization*, i.e. when the first set-up of your repository.

The simplest way to re-sign a metadata file is using the command line interface: type `tufup sign --help` for options

Basically, you can do something like this:

```shell
tufup sign -e 3 timestamp /path/to/my/private/keys
```
where `-e` sets the number of days until expiration.

To understand what this command does, have a look at [repo/cli.py][3].

Alternatively, if you would like to re-sign using a python script, have a look at [repo_workflow_example.py][1], which shows how to perform some common repository tasks, for example:

```python
...

# Initialize repo from config
repo = Repository.from_config()

# Re-sign expired roles (downstream roles are refreshed automatically)
repo.refresh_expiration_date(role_name='timestamp', days=3)
repo.publish_changes(private_key_dirs=['/path/to/my/private/keys'])

...
```

Note: If you need to re-sign the `snapshot` metadata, that will automatically re-sign the `timestamp` as well.

> Second: with your answer i am coming up with one more doubt like suppose in the client machine where application is running and it has root.json file along with it [our main.spec ensures that the latest root.json metadata file is included in the bundle (given in example)] and when we re-sign the json file , eventually/obviously the changes will be made on the server side tufup repository's metadata files, then how would the client side signature will match the changed signature when the Client downloads the metadata over the client machine, it will eventually rejects the new archive and shows Error.

Not sure I completely understand your question, but, there is a chain of trust starting with the root.json that is included in your initial application bundle. If the root.json is re-signed on the server side, the client will pick this up automatically.

The best way to understand this is by studying the [detailed client workflow][2] in the TUF specification. This explains in detail how the client handles metadata changes.

[1]: https://github.com/dennisvang/tufup/blob/master/examples/repo/repo_workflow_example.py#L183
[2]: https://theupdateframework.github.io/specification/latest/#detailed-client-workflow
[3]: https://github.com/dennisvang/tufup/blob/9e53b744a8811d0d9ab6820172da11c9a4327453/src/tufup/repo/cli.py#L320
